kvdb/etcd/stm.go

```diff
@@ -753,6 +969,72 @@ func (s *stm) OnCommit(cb func()) {
 	s.onCommit = cb
 }
 
+// Prefetch will prefetch the passed keys and prefixes in one transaction.
+// Keys and prefixes that we already have will be skipped.
+func (s *stm) Prefetch(keys []string, prefixes []string) {
+	fetchKeys := make([]string, 0, len(keys))
+	for _, key := range keys {
+		if _, ok := s.rset.getItem(key); !ok {
+			fetchKeys = append(fetchKeys, key)
+		}
+	}
+
+	fetchPrefixes := make([]string, 0, len(prefixes))
+	for _, prefix := range prefixes {
+		if s.rset.hasFullRange(prefix) {
```

Comment:
Since the read set has just been created yet and we haven't pre-fetched yet, won't this always be false? In that we wouldn't have pulled in any of these keys?

Reply:
This is to make sure that if the user calls `Prefetch` with already prefetched keys/ranges in the same transaction then we won't fetch them again. One such example is if we retry.